ibis/backends/duckdb/__init__.py

```diff
@@ -0,0 +1,146 @@
+"""DuckDB backend."""
+from __future__ import annotations
+
+import itertools
+from typing import Mapping
+
+import sqlalchemy as sa
+
+import duckdb
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.schema as sch
+import ibis.expr.types as ir
+from ibis.backends.base.sql import BaseSQLBackend
+
+from ..base.sql.alchemy.database import AlchemyTable as _AlchemyTable
+from ..base.sql.alchemy.datatypes import to_sqla_type
+from .compiler import DuckDBSQLCompiler
+
+
+# TODO: AlchemyTable calls infer on the sqla_table attribute in the super call
+# in DuckDBTable -- it wants a sa.Table this is a hack to get it to make it
+# through the parent __init__ without barfing
+@sch.infer.register(sa.sql.selectable.TableClause)
+def schema_passthrough(table, schema=None):
+    return schema
+
+
+class DuckDBTable(_AlchemyTable):
+    """
+    References a table in a DuckDB file / memory
+
+    We construct a sqlalchemy-compatible table at creation time to allow using
+    the postgres compiler without recursing infinitely.
+
+    See base.sql.alchemy.query_builder._AlchemyTableSetFormatter._format_table
+    """
+
+    def __init__(self, name, schema, source):
+        sqla_table = sa.table(
+            name,
+            *(sa.column(n, to_sqla_type(t)) for n, t in schema.items()),
+        )
+        super().__init__(sqla_table, source, schema)
+
+
+class Backend(BaseSQLBackend):
+    name = "duckdb"
+    compiler = DuckDBSQLCompiler
+    table_class = DuckDBTable
+
+    def current_database(self):
+        raise NotImplementedError
+
+    @property
+    def version(self) -> str:
+        # TODO: there is a `PRAGMA version` we could use instead
+        try:
+            import importlib.metadata as importlib_metadata
+        except ImportError:
+            # TODO: remove this when Python 3.9 support is dropped
+            import importlib_metadata
+        return importlib_metadata.version("duckdb")
+
+    def do_connect(
+        self, path: str | None = None, read_only: bool = False
+    ) -> None:
+        """Create an Ibis client connected to a DuckDB database."""
+
+        if path is None:
+            path = ":memory:"
```

Comment:
I think we can just set the default value of `path` to `":memory:"` and type the `path` parameter as `path: str`.